crates/sargon/src/system/sargon_os/pre_authorization/sargon_os_pre_authorization_status.rs

```diff
@@ -0,0 +1,360 @@
+                    0 // Avoid overflow in case we check after expiration
+                }
+            }
+            DappToWalletInteractionSubintentExpiration::AfterDelay(delay) => {
+                delay.expire_after_seconds
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod poll_pre_authorization_status_with_delays {
+    use super::*;
+    use actix_rt::time::timeout;
+    use std::{future::Future, time::Duration};
+
+    #[allow(clippy::upper_case_acronyms)]
+    type SUT = SargonOS;
+    #[allow(clippy::upper_case_acronyms)]
+    type SSR = SubintentStatusResponse;
+
+    #[actix_rt::test]
+    async fn success_on_third_poll() {
+        // This test will simulate the case where the first two polls return `Unknown`,
+        // while the third one returns `CommittedSuccess`.
+        let result = simulate_poll_status(
+            2,
+            SSR::sample_committed_success(),
+            expiration_after_delay(10),
+        )
+        .await
+        .unwrap();
+
+        // The result is Success with the mocked TransactionIntentHash
+        assert_eq!(
+            result.0,
+            PreAuthorizationStatus::Success(
+                TransactionIntentHash::sample().to_string()
+            )
+        );
+
+        // and there should have been a delay of 2s after first call, and 3s after the second call
+        assert_eq!(result.1, vec![2, 3]);
+    }
+
+    #[actix_rt::test]
+    async fn success_exactly_at_expiration() {
+        // This test will simulate the case where the subintent expires in 9 seconds, and we only get
+        // the `CommittedSuccess` on the fourth request (after 9 seconds of delay).
+        let result = simulate_poll_status(
+            3,
+            SSR::sample_committed_success(),
+            expiration_after_delay(9),
+        )
+        .await
+        .unwrap();
+
+        // The result is Success with the mocked TransactionIntentHash
+        assert_eq!(
+            result.0,
+            PreAuthorizationStatus::Success(
+                TransactionIntentHash::sample().to_string()
+            )
+        );
+
+        // and delays should have been incrementing as expected until the last one (which is reduced to the remaining time)
+        assert_eq!(result.1, vec![2, 3, 4]);
+    }
+
+    #[actix_rt::test]
+    async fn success_immediately_after_expiration() {
+        // This test will simulate the case where the subintent expires in 10 seconds, and we only get
+        // the `CommittedSuccess` on the fifth request (after 11 seconds of delay).
+        let result = simulate_poll_status(
+            4,
+            SSR::sample_committed_success(),
+            expiration_after_delay(10),
+        )
+        .await
+        .unwrap();
+
+        // The result is Success with the mocked TransactionIntentHash
+        assert_eq!(
+            result.0,
+            PreAuthorizationStatus::Success(
+                TransactionIntentHash::sample().to_string()
```

Comment:
I like asserting exact values like this. Just wanna present an alternative solution, which you might or might not know if.

If it would be hard to know the exact expected value of the IntentHash, you can use the `matches!` macro, which is really really great in situation like this! You would then write like this:

```rust
assert!( // note, not using assert_eq! macro
    matches!(
        result.0,
        PreAuthoriziationStatus::Success(_)
    )
);
```

Reply:
right, I had used the `matches!` macro at other places where we don't care/know about the associated value. Makes sense here since the associated value is actually set on mocks (and no need to test our mocks)